core/lib/contract_verifier/src/resolver/github/gh_api.rs

```diff
@@ -0,0 +1,240 @@
+                .set_connect_timeout(Some(Self::connect_timeout()))
+                .set_read_timeout(Some(Self::read_timeout()))
+                .build()
+                .unwrap(),
+        );
+        Self { client }
+    }
+
+    fn retry_config() -> RetryConfig {
+        RetryConfig::Simple(4)
+    }
+
+    fn connect_timeout() -> Duration {
+        Duration::from_secs(10)
+    }
+
+    fn read_timeout() -> Duration {
+        Duration::from_secs(60)
+    }
+
+    /// Returns versions for both upstream and our fork of solc.
+    pub async fn solc_versions(&self) -> anyhow::Result<HashMap<String, reqwest::Url>> {
+        let mut versions = self
+            .extract_versions(CompilerGitHubRelease::Solc)
+            .await
+            .context("Can't fetch upstream solc versions")?;
+        versions.extend(
+            self.extract_versions(CompilerGitHubRelease::ZkVmSolc)
+                .await
+                .context("Can't fetch zkVM solc versions")?,
+        );
+        Ok(versions)
+    }
+
+    pub async fn zksolc_versions(&self) -> anyhow::Result<HashMap<String, reqwest::Url>> {
+        self.extract_versions(CompilerGitHubRelease::ZkSolc).await
+    }
+
+    pub async fn vyper_versions(&self) -> anyhow::Result<HashMap<String, reqwest::Url>> {
+        self.extract_versions(CompilerGitHubRelease::Vyper).await
+    }
+
+    pub async fn zkvyper_versions(&self) -> anyhow::Result<HashMap<String, reqwest::Url>> {
+        self.extract_versions(CompilerGitHubRelease::ZkVyper).await
+    }
+
+    /// Will scan all the releases for a specific compiler.
+    async fn extract_versions(
+        &self,
+        compiler: CompilerGitHubRelease,
+    ) -> anyhow::Result<HashMap<String, reqwest::Url>> {
+        // Create a stream over all the versions to not worry about pagination.
+        let releases = self
+            .client
+            .repos(compiler.organization(), compiler.repo())
+            .releases()
+            .list()
+            .per_page(100)
+            .send()
+            .await?
+            .into_stream(&self.client);
+        tokio::pin!(releases);
+
+        // Go through all the releases, filter ones that match the version.
+        // For matching versions, find a suitable asset and store its URL.
+        let mut versions = HashMap::new();
+        while let Some(release) = releases.try_next().await? {
+            // Skip pre-releases.
+            if release.prerelease {
+                continue;
+            }
+
+            if let Some(version) = compiler.extract_version(&release.tag_name) {
+                if compiler.is_version_blacklisted(&version) {
+                    tracing::debug!("Skipping {compiler:?}:{version} due to blacklist");
+                    continue;
+                }
+
+                let mut found = false;
+                for asset in release.assets {
+                    if compiler.match_asset(asset.browser_download_url.as_str()) {
```

Comment:
From what I understood `url` would look like `/releases/assets/1`, while `browser_download_url` would look like `/releases/download/v1.0.0/my-awesome-artifact` (see [API docs](https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#list-releases)). If we are to use `url`, then we need to match asset based on the `name` field, but since both approaches work, I simply chose the simpler one.